[Michael Low] iP

michaelseyo opened the pull request:
## BerniePro
> “Appear weak when you are strong, and strong when you are weak.”
― Sun Tzu, The Art of War

BerniePro remembers stuff you can't remember. It's,
- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

All you have to do is,
1. download it from [here](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html)
2. double-tap that
3. add your tasks
4. let it manage your life for you :wink:

And it is **FREE**!

Features:
[x] Managing tasks
[ ] Managing deadlines (coming soon)
[ ] Reminders (coming soon)

If you're a Java programmer, you can use it to practice Java too! Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```
Requirements:
~a heading~
~a bullet list~
~a numbered list~
~a fenced code block (with syntax highlighting)~
~a task list~
~an emoji~
~a blockquote~
~a hyper link~
~inline code~
~some text formatting: bold, italic, stikethrough etc.~

Commit:
Add Gradle support

michaelseyo committed:
level-1

michaelseyo committed:
level-2

michaelseyo committed:
marking logic implemented

michaelseyo committed:
implemented Task class

michaelseyo committed:
integrated Task class into program

michaelseyo committed:
added new classes to represent differen Task

michaelseyo committed:
fixed bug for list action

michaelseyo committed:
integrated Task subclasses into program function

michaelseyo committed:
updated expected and input

michaelseyo committed:
updated Exception handling

michaelseyo committed:
Added linebreaks and updated text ui testing

michaelseyo committed:
Exception handling

michaelseyo committed:
updated javadocs

michaelseyo committed:
Added delete fn

michaelseyo committed:
Added enums

michaelseyo committed:
create save and read functionality

michaelseyo committed:
create basic level-8 with check for yyyy-mm-dd format

michaelseyo committed:
Merge branch 'branch-Level-7'

michaelseyo committed:
fix conflicts

michaelseyo committed:
add Ui class with print methods

michaelseyo committed:
integrate Ui Class into Bernie class

michaelseyo committed:
add Storage class

michaelseyo committed:
create parser part1

michaelseyo committed:
create parser class part2

michaelseyo committed:
change enum to be separate class

michaelseyo committed:
organise classes into packages

michaelseyo committed:
add jUnit tests

michaelseyo committed:
create jar

michaelseyo committed:
Updated javadocs

michaelseyo committed:
Tweak to coding standards

michaelseyo committed:
Add find function to find tasks by words

michaelseyo committed:
Fix conflicts between A-JavaDocs and A-CodingStandards

michaelseyo committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find function to find tasks by words

michaelseyo committed:
Fix bugs after merge

michaelseyo committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

* origin/add-gradle-support:
  Add Gradle support

michaelseyo committed:
Add graddle support with unit testing

michaelseyo committed:
Use CheckStyle and fix style errors

yuqitanyq commented on src/main/java/bernie/commands/AddCommand.java:
```diff
@@ -0,0 +1,74 @@
+        } else {
+            throw new BernieException("Not a valid type of task!");
+        }
+    }
```
I like how there are no nested blocks of if-else statements!

michaelseyo replied:
hi yuqi thanks for the overall feedback, will take it into account!

yuqitanyq commented on src/main/java/bernie/storage/Storage.java:
```diff
@@ -0,0 +1,103 @@
+                // create dir and file
+                dataDir.mkdir();
+                tasksFile.createNewFile();
+            }
```
Not sure if it's possible, but maybe abstracting out the while loop into a method might improve readability!

yuqitanyq commented on src/main/java/bernie/tasks/TaskList.java:
```diff
@@ -0,0 +1,157 @@
+            }
+        }
+        return count;
+    }
```
I like how the code is modularised and abstracted out so that it is easy to read!

yuqitanyq commented on src/main/java/bernie/enums/Type.java:
```diff
@@ -0,0 +1,22 @@
+    UNMARK,
+    DELETE,
+    FIND
+}
```
I like how the enum constants are declared to be simple and easy to understand.

yuqitanyq commented on src/main/java/bernie/parser/Parser.java:
```diff
@@ -0,0 +1,221 @@
+    public String[] getParams(Type taskType, String input) throws InvalidArgumentException {
+        String description;
+        String[] parsedArr = null;
+        switch (taskType) {
+        case TODO:
+            parsedArr = input.split("todo ");
+            description = getDescription(parsedArr, Type.TODO);
+            parsedArr = new String[]{ description };
+            break;
+        case DEADLINE:
+            parsedArr = input.split(" /by ");
+            description = getDescription(parsedArr, Type.DEADLINE);
+            String by = getTime(parsedArr, Type.DEADLINE);
+            parsedArr = new String[]{ description, by };
+            break;
+        case EVENT:
+            parsedArr = input.split(" /at ");
+            description = getDescription(parsedArr, Type.EVENT);
+            String at = getTime(parsedArr, Type.EVENT);
+            parsedArr = new String[]{ description, at };
+            break;
+        case MARK:
+            parsedArr = input.split(" ");
+            // check valid
+            checkMarkOrDeleteInput(parsedArr, Type.MARK);
+            checkValidMarkAction(parsedArr);
+            break;
+        case DELETE:
+            parsedArr = input.split(" ");
+            checkMarkOrDeleteInput(parsedArr, Type.DELETE);
+            parsedArr = new String[]{ parsedArr[1] };
+            break;
+        case FIND:
+            parsedArr = input.split("find ");
+            description = getDescription(parsedArr, Type.FIND);
+            parsedArr = new String[]{ description };
+            break;
```
Not sure if it's possible, but maybe abstracting out the splitting of string might improve readability!

yuqitanyq commented on src/main/java/bernie/tasks/TaskList.java:
```diff
@@ -0,0 +1,157 @@
+     * @param taskType Type, either TODO, DEADLINE or EVENT
+     * @return Task created
+     */
+    public Task addTask(String[] parsedArr, Type taskType) {
+        Task newTask = null;
+        String description;
+        LocalDate by;
+        String at;
+        switch (taskType) {
+        case TODO:
+            description = parsedArr[0];
+            newTask = new ToDo(description);
+            break;
+        case DEADLINE:
+            description = parsedArr[0];
+            by = LocalDate.parse(parsedArr[1]);
+            newTask = new Deadline(description, by);
+            break;
+        case EVENT:
+            description = parsedArr[0];
+            at = parsedArr[1];
+            newTask = new Event(description, at);
+            break;
+        default:
+            break;
+        }
+        tasks.add(newTask);
+        return newTask;
+    }
```
I like how easy it is to read this method! 👍

yuqitanyq reviewed (commented):
Overall, good job on the code quality! LGTM, except for some minor suggestions on abstractions!

michaelseyo committed:
Create basic gui without loadtasks and greet on start

vishandi commented on src/main/java/bernie/Bernie.java:
```diff
@@ -0,0 +1,18 @@
+ */
+public class Bernie {
+    /**
+     * Runs Bernie to help you remember your tasks. CommandHandelr
```
A minor typo here. CommandHandelr -> CommandHandler :)

vishandi commented on src/main/java/bernie/commands/AddCommand.java:
```diff
@@ -0,0 +1,74 @@
+     * the message to the user.
+     */
+    public void execute() {
+        Task newTask = null;
```
This is fine, but you could consider declaring `Task newTask;` without `= null`.

michaelseyo replied:
Hi vishandi, tried to do that but seems like i'll get an error that i might not initialize the Task, but thanks for the feedback i've worked on some of the things you mentioned

vishandi commented on src/main/java/bernie/commands/CommandHandler.java:
```diff
@@ -0,0 +1,96 @@
+     * @param input String, user input
+     * @return boolean value, indicating if the program will end or not.
+     */
+    public boolean handle (String input) {
```
I like the functionality of this method, but as per Java coding standard, you could consider changing the method name or changing the return type. Also, a minor typo in which there shouldn't be a space between `handle` and `(String input)`.

vishandi commented on src/main/java/bernie/commands/MarkCommand.java:
```diff
@@ -0,0 +1,51 @@
+        String taskNum = parsedArr[1];
+        TaskList tasks = getTasks();
+        UiHandler uiHandler = getUiHandler();
+        if (action.equals("mark")) {
+            Task markedTask = tasks.markTask(Type.MARK, taskNum);
+            uiHandler.showDoneMsg(markedTask);
+        } else if (action.equals("unmark")) {
+            Task unmarkedTask = tasks.markTask(Type.UNMARK, taskNum);
+            uiHandler.showUndoneMsg(unmarkedTask);
+        }
```
One suggestion from me is changing this if-else statements to switch-case statement to increase readability.

vishandi commented on src/main/java/bernie/exceptions/InvalidArgumentException.java:
```diff
@@ -0,0 +1,11 @@
+ * InvalidArgumentException is thrown whenever the user enters invalid arguments into a valid command. Such examples
+ * are marking a task when it is already marked, or marking a task that doesn't exist, etc.
+ */
+public class InvalidArgumentException extends Exception {
+    public InvalidArgumentException(String errorMessage) {
+        super(errorMessage);
+    }
+}
```
Instead of inheriting from `Exception` class, you could consider inheriting from `BernieException`. Moreover, as you add more exceptions to this package, you could also consider making all the "handmade" exception as static attribute to `BernieException` class.

vishandi commented on src/main/java/bernie/parser/Parser.java:
```diff
@@ -0,0 +1,221 @@
+     * @param parsedArr
+     * @throws InvalidArgumentException if user attempts to marks a marked task or unmark an unmarked task
+     */
+    void checkValidMarkAction(String[] parsedArr) throws InvalidArgumentException {
```
Just checking, are the methods beyond this line supposed to be `public`?

vishandi commented on src/main/java/bernie/Bernie.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * Bernie, your personal tasks manager! :)
+ */
+public class Bernie {
```
Nice name btw!

vishandi reviewed (commented):
I really like the JavaDocs in your codes. They are very well written and descriptive. I think the parts that you could improve on are the method names and the accessibility of the methods. Overall, I really like the structure of your code! :)

michaelseyo committed:
Create gui with styling and update loading task logic

michaelseyo committed:
Add var-args for inputResponder fn

michaelseyo committed:
Style checks and re-organise some code

michaelseyo committed:
Add assertions to various parts of the code

Code assumes that program will have certain states

The program may not run as we think it would and could result in unexpected states

Let's add assertions so that we can assert that the program is in the current state. We add assertions mainly in areas where we assert that a task has been created and not null, tasks that we get are not null and also asserting of no negative values for taskList size

Using assertions this way allows us to assert that parts of the proram that is not determine by the user inputs are running correctly

michaelseyo committed:
Merge pull request #1 from michaelseyo/branch-A-Assertions

Add assertions to various parts of the code

michaelseyo committed:
Improve code quality

Some code has magic numbers/strings and organisation is messy

The unclear code could result in a hard time in understanding the code

Guard clauses, local variables, refactoring, extracting and abstracting methods were done to help improve code quality

This ensures that the code has a smooth flow to it and is easy to understand

michaelseyo committed:
Add streams for getting num tasks left not done

michaelseyo committed:
Merge pull request #2 from michaelseyo/branch-A-CodeQuality

Improve code quality

michaelseyo committed:
Fix merge conflicts

michaelseyo committed:
Merge pull request #3 from michaelseyo/branch-A-Streams

Branch a streams

michaelseyo committed:
Configure CI

michaelseyo committed:
Revert "Configure CI"

This reverts commit 14d671adf783b56d1c754ad23bc78885adae9794.

michaelseyo committed:
Create gradle.yml

michaelseyo committed:
Merge pull request #4 from michaelseyo/branch-A-CI

Create CI

michaelseyo committed:
Create reminders feature

michaelseyo committed:
Add different colour textbox for user and bernie

michaelseyo committed:
Clean up code for setting diff colour textbox

michaelseyo committed:
Add background

michaelseyo committed:
Update styling and add help button

michaelseyo committed:
Add user guide

michaelseyo committed:
Fix padding for textfield

michaelseyo committed:
Add check for description with white space only

michaelseyo committed:
Fix style and ui changes

michaelseyo committed:
Rename README.md to README .md
